Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -145,8 +145,217 @@ public struct HTTPResponseHead: Equatable {
     }
 }
 
-fileprivate typealias HTTPHeadersStorage = [String: [(String, String)]] // [lowerCasedName: [(originalCaseName, value)]
+/// The Index for a header name or value that points into the underlying `ByteBuffer`.
+typealias HTTPHeaderIndex = (start: Int, length: Int)
+
+/// Internal storage used for `HTTPHeaders`.
+private final class HTTPHeadersByteStorage : CustomStringConvertible {
+    private var buffer: ByteBuffer
+    fileprivate var indicies: [HTTPHeaderIndex]
+
+    init(buffer: ByteBuffer, indicies: [HTTPHeaderIndex]) {
+        assert(indicies.count % 2 == 0)
+        self.buffer = buffer
+        self.indicies = indicies
+    }
+
+    /// Returns `true` if no headers are stored.
+    var isEmpty: Bool {
+        return self.indicies.isEmpty
+    }
+
+    /// Returns the number of headers stored.
+    var count: Int {
+        return self.indicies.count
+    }
+
+    /// Find the first `HTTPHeaderIndex` of the value for the name.
+    ///
+    /// - parameters:
+    ///     - name: The name of the header.
+    /// - returns: The index of the value for the given name.
+    func findFirst(name: String) -> HTTPHeaderIndex? {
+        guard !self.indicies.isEmpty else {
+            return nil
+        }
+
+        let utf8 = name.utf8
+        let stringLength = utf8.count
+        var idx = 0
+        while idx < self.indicies.count {
```

Comment:
This is probably more naturally expressed as `for index in stride(from: 0, to: self.indices.count, by: 2)`